dirty_cat/gamma_poisson_factorization.py

```diff
@@ -0,0 +1,600 @@
+        self.init = init
+        self.analyzer = analyzer
+        self.add_words = add_words
+        self.random_state = check_random_state(random_state)
+        self.rescale_W = rescale_W
+        self.max_iter_e_step = max_iter_e_step
+
+        # Init n-grams counts vectorizer
+        if self.hashing:
+            self.ngrams_count = HashingVectorizer(
+                 analyzer=self.analyzer, ngram_range=self.ngram_range,
+                 n_features=self.hashing_n_features,
+                 norm=None, alternate_sign=False)
+            if self.add_words: # Init a word counts vectorizer if needed
+                self.word_count = HashingVectorizer(
+                     analyzer='word',
+                     n_features=self.hashing_n_features,
+                     norm=None, alternate_sign=False)
+        else:
+            self.ngrams_count = CountVectorizer(
+                 analyzer=self.analyzer, ngram_range=self.ngram_range)
+            if self.add_words:
+                self.word_count = CountVectorizer()
+
+    def _update_H_dict(self, X, H):
+        """
+        For each category x in X, update the dictionary self.H_dict with
+        the corresponding bag-of-n-grams representation h.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, )
+            The string data to fit the model on.
+        H : array-like, shape (n_samples, n_vocab)
+            The corresponding bag-of-n-grams representations.
+        """
+        for x, h in zip(X, H):
+            self.H_dict[x] = h
+
+    def _init_vars(self, X):
+        """
+        Build the bag-of-n-grams representation V of X and initialize
+        the topics W.
+        """
+        # Build the n-grams counts matrix unq_V on unique elements of X
+        unq_X, lookup = np.unique(X, return_inverse=True)
+        unq_V = self.ngrams_count.fit_transform(unq_X)
+        if self.add_words: # Add word counts to unq_V
+            unq_V2 = self.word_count.fit_transform(unq_X)
+            unq_V = sparse.hstack((unq_V, unq_V2), format='csr')
+
+        if not self.hashing: # Build n-grams/word vocabulary
+            self.vocabulary = self.ngrams_count.get_feature_names()
+            if self.add_words:
+                self.vocabulary = np.concatenate(
+                    (self.vocabulary, self.word_count.get_feature_names()))
+
+        _, self.n_vocab = unq_V.shape
+        # Init the topics W given the n-grams counts V
+        self.W_, self.A_, self.B_ = self._init_w(unq_V[lookup], X)
+        # Init the activations unq_H of each unique input string
+        unq_H = _rescale_h(unq_V, np.ones((len(unq_X), self.n_topics)))
+        # Map unique input strings to their activations
+        self.H_dict = dict()
+        self._update_H_dict(unq_X, unq_H)
+        if self.rescale_rho:
+            # Make update rate per iteration independant of the batch_size
+            self.rho_ = self.rho ** (self.batch_size / len(X))
+        else:
```

Comment:
I guess we can do away with the else block

Reply:
Yes, I moved `self.rho_ = self.rho` in the init method.

Reply:
No, it should not be in the init, because if people change rho or batch_size before calling fit, the change won't be propagated. I guess what @amy12xx meant is that you can do:
```python
self.rho_ = rho
if self.rescale_rho:
       # Make update rate per iteration independant of the batch_size
       self.rho_ = self.rho ** (self.batch_size / len(X))
```